File: Difficulty: Medium/Sort a linked list of 0s, 1s and 2s/sort-a-linked-list-of-0s-1s-and-2s.java
/*
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}
*/

class Solution {
    static Node segregate(Node head) {

        int zero = 0;
        int one = 0;
        int two = 0;

        Node temp = head;

        while (temp != null) {

            if (temp.data == 0)
                zero++;
            else if (temp.data == 1)
                one++;
            else
                two++;

            temp = temp.next;
        }

        temp = head;

        while (zero > 0) {
            temp.data = 0;
            zero--;
            temp = temp.next;
        }

        while (one > 0) {
            temp.data = 1;
            one--;
            temp = temp.next;
        }

        while (two > 0) {
            temp.data = 2;
            two--;
            temp = temp.next;
        }

        return head;
    }
}

File: Difficulty: Medium/Sort a linked list of 0s, 1s and 2s/sort-a-linked-list-of-0s-1s-and-2s.cpp
class Solution {
  public:
    Node* segregate(Node* head) {
        if (head == nullptr || head->next == nullptr)
            return head;

        Node zeroDummy(0);
        Node oneDummy(0);
        Node twoDummy(0);

        Node* zero = &zeroDummy;
        Node* one = &oneDummy;
        Node* two = &twoDummy;

        Node* curr = head;

        while (curr != nullptr) {
            if (curr->data == 0) {
                zero->next = curr;
                zero = zero->next;
            }
            else if (curr->data == 1) {
                one->next = curr;
                one = one->next;
            }
            else {
                two->next = curr;
                two = two->next;
            }

            curr = curr->next;
        }

        // 2 list ka end
        two->next = nullptr;

        // 0 -> 1 -> 2
        zero->next = (oneDummy.next != nullptr) ? oneDummy.next : twoDummy.next;
        one->next = twoDummy.next;

        return zeroDummy.next;
    }
};